sunpy/net/helioviewer.py

```diff
@@ -171,25 +166,79 @@ def download_jp2(self, date, observatory=None, instrument=None, detector=None,
         >>> aia = sunpy.map.Map(filepath)  # doctest: +REMOTE_DATA
         >>> aia.peek()  # doctest: +SKIP
         """
-        if source_id is None:
+        key = (observatory, instrument, detector, measurement)
+        source_id = self._get_source_id(source_id, key)
+
+        params = {
+            "action": "getJP2Image",
+            "date": self._format_date(date),
+            "sourceId": source_id,
+        }
+
+        return self._get_file(params, directory=directory, overwrite=overwrite)
+
+    def get_jp2_header(self, date, observatory=None, instrument=None, detector=None, measurement=None, jp2_id=None):
+        """
+        Get the XML header embedded in a JPEG2000 image. Includes the FITS header as well as a section
+        of Helioviewer-specific metadata.
+
+        This uses `getJP2Header <https://api.helioviewer.org/docs/v2/#JPEG2000>`_ from the Helioviewer API.
+
+        Parameters
+        ----------
+        date : `astropy.time.Time`, `str`
+            A `~sunpy.time.parse_time` parsable string or `~astropy.time.Time`
+            object for the desired date of the image
+        observatory : `str`
+            Observatory name
+        instrument : `str`
+            Instrument name
+        measurement : `str`
+            Measurement name
+        detector : `str`
+            Detector name
+        jp2_id : `int`
+            Unique JP2 image identifier.
+            This can be used directly instead of using the previous parameters.
+
+        Returns
+        -------
+        out : `dict`
+            Returns a dictionary containing the header information of JPEG 2000 image.
+            The returned dictionary may have either one or two keys: *fits* and *helioviewer*.
+
+        Examples
+        --------
+        >>> from sunpy.net import helioviewer
+        >>> hv = helioviewer.HelioviewerClient()  # doctest: +REMOTE_DATA
+        >>> header = hv.get_jp2_header('2012/07/03', observatory='SDO',
+        ...                            instrument='HMI', detector=None, measurement='continuum')  # doctest: +REMOTE_DATA
+        >>> # The key 'fits' can be used to get the fits header information
+        >>> fits_header = header['fits']  # doctest: +REMOTE_DATA
+        >>> # The keys 'helioviewer' can be used to extract the helioviewer specific metadata.
+        >>> helioviewer_meta_data = header['helioviewer']  # doctest: +REMOTE_DATA
+        """
+        if jp2_id is None:
             try:
-                key = (observatory, instrument, detector, measurement)
-                source_id = self.data_sources[key]
+                jp2_id = self.get_closest_image(date, observatory, instrument, detector, measurement)['id']
             except KeyError:
                 raise KeyError("The values used for observatory, instrument, detector, measurement "
                                "do not correspond to a source_id. Please check the list using "
```

Comment:
Do we need the `KeyError` here now?

Reply:
Yes we need it because this function does not use source_id as a parameter but as a way to get jp2_id. But we need to check if the user has entered the correct source id

Reply:
But the error now gets raised from inside `get_closest_image`?

Reply:
Oh, Ok I got it. Yes I will remove it